units/src/fee.rs

```diff
@@ -143,27 +144,47 @@ impl FeeRate {
         // No `?` operator in const context.
         match self.to_sat_per_kwu().checked_mul(weight.to_wu()) {
             Some(mul_res) => match mul_res.checked_add(999) {
-                Some(add_res) => Some(Amount::from_sat(add_res / 1000)),
+                // Unchecked because div by 1000 cannot be out of range.
+                Some(add_res) => Some(Amount::from_sat_unchecked(add_res / 1000)),
                 None => None,
             },
             None => None,
         }
     }
 }
 
-/// Computes the ceiling so that the fee computation is conservative.
 impl ops::Mul<FeeRate> for Weight {
     type Output = Amount;
 
+    /// Enables `fee = fee_rate * weight`.
+    ///
+    /// Computes the ceiling so that the fee computation is conservative. Consider using
+    /// [`FeeRate::checked_mul_by_weight`] if you do not control both the fee rate and the weight
+    /// because this function can panic.
+    ///
+    /// # Panics
+    ///
+    /// If `fee_rate * weight` exceeds `Amount::MAX_MONEY`.
     fn mul(self, rhs: FeeRate) -> Self::Output {
-        Amount::from_sat((rhs.to_sat_per_kwu() * self.to_wu() + 999) / 1000)
+        rhs.checked_mul_by_weight(self).expect("attempted to calculate absurdly high fee")
```

Comment:
Please consider acking this change even though it is not the final state of this function. To correctly handle the error case when some math operation exceeds the invariant we need to do a bunch more work. This PR has repeatedly been nacked because it contains too many changes. Note the `amount` module is totally f'ed right now. We are just trying to make forward progress. It is _really_ dragging on. Please be considerate of my sanity when reviewing.

Reply:
Perhaps a followup issue then so someone else can tackle it and so it's not forgotten since you found it important enough to mention here.